rust/cubesql/cubesql/src/compile/engine/udf/common.rs

```diff
@@ -2313,6 +2313,78 @@ macro_rules! generate_series_udtf {
     }};
 }
 
+macro_rules! generate_series_non_primitive_udtf {
+    ($ARGS:expr, $TYPE: ident, $PRIMITIVE_TYPE: ident) => {{
+        let mut section_sizes: Vec<usize> = Vec::new();
+        let l_arr = &$ARGS[0].as_any().downcast_ref::<PrimitiveArray<$TYPE>>();
+
+        if l_arr.is_some() {
+            let l_arr = l_arr.unwrap();
+            let r_arr = downcast_primitive_arg!($ARGS[1], "right", $TYPE);
+            let step_arr = IntervalMonthDayNanoArray::from_value(
+                IntervalMonthDayNanoType::make_value(0, 1, 0), // 1 day as default
+                1,
+            );
+            let step_arr = if $ARGS.len() > 2 {
+                downcast_primitive_arg!($ARGS[2], "step", IntervalMonthDayNanoType)
+            } else {
+                &step_arr
+            };
+
+            let mut builder = PrimitiveBuilder::<$TYPE>::new(1);
+            for (i, (start, end)) in l_arr.iter().zip(r_arr.iter()).enumerate() {
+                let step = if step_arr.len() > i {
+                    step_arr.value(i)
+                } else {
+                    step_arr.value(0)
+                };
+
+                if let (Some(start), Some(end)) = (start, end) {
+                    let mut section_size: i64 = 0;
+                    if start <= end && step > 0 {
+                        let mut current = start;
+                        loop {
+                            if current > end {
+                                break;
+                            }
+                            builder.append_value(current).unwrap();
+
+                            section_size += 1;
+
+                            if $ARGS[0].as_any().downcast_ref::<Date32Array>().is_some() {
```

Comment:
This macro will expand 3 times. These checks will be in all the codes. it's better to pass block as macros's argument in the calls.

Reply:
Yeah! Good idea! Thnx!

Reply:
Rewrote with few macros without if-else-checks.